### PR TITLE
fix: sso users not able to delete clients

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

the password was always required for deleting a client

### Solutions

try to delete/register without a password -> if BE response with missing auth -> show dialog and delete?register with password

Needs releases with:

- [ ] GitHub link to other pull request

### Testing

WIP

- [x] I have added automated test to this contribution

#### How to Test

login with SSO -> too many clients screen -> delete one -> alles gut

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/2/) **failed**.

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/4/) **failed**.

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/5/) **failed**.

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/6/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceViewModel.kt [hunk 1]
```diff
@@ -78,16 +71,64 @@ class RemoveDeviceViewModel @Inject constructor(
...
-    fun onItemClicked(device: Device) {
-        updateStateIfSuccess { it.copy(removeDeviceDialogState = RemoveDeviceDialogState.Visible(device = device)) }
+    fun onItemClicked(client: Client) {
+        tryToDeleteOrShowPasswordDialog(client)
+    }
+
+    private fun tryToDeleteOrShowPasswordDialog(client: Client) {
+        // try delete with no password (will success only for SSO accounts)
+        viewModelScope.launch(Dispatchers.Main) {
+            // try delete with no password (will success only for SSO accounts
```
Comment (saleniuk) [inline app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceViewModel.kt:81]: Duplicated comment 😄

File: app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceViewModel.kt [hunk 2]
```diff
@@ -78,16 +71,64 @@ class RemoveDeviceViewModel @Inject constructor(
...
+                                removeDeviceDialogState = RemoveDeviceDialogState.Visible(
+                                    client = client
+                                )
+                            )
+                        }
+                        is RegisterClientResult.Failure.Generic -> state = RemoveDeviceState.Error(result.genericFailure)
+                        RegisterClientResult.Failure.InvalidCredentials -> {
+                            // server will never response with InvalidCredentials in case of trying with no password
+                        }
+                        RegisterClientResult.Failure.TooManyClients -> {}
```
Comment (saleniuk) [inline app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceViewModel.kt:100]: Maybe it should also have some informative comment which explains why it won't happen

File: app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceViewModel.kt [hunk 3]
```diff
@@ -78,16 +71,64 @@ class RemoveDeviceViewModel @Inject constructor(
...
+                    }
+                }
+                is DeleteClientResult.Failure.Generic -> state = RemoveDeviceState.Error(deleteResult.genericFailure)
+                DeleteClientResult.Failure.InvalidCredentials -> updateStateIfSuccess {
+                    it.copy(
+                        removeDeviceDialogState = RemoveDeviceDialogState.Visible(
+                            client = client
+                        )
+                    )
+                }
```
Comment (saleniuk) [inline app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceViewModel.kt:114]:
```
updateStateIfSuccess {
                    it.copy(
                        removeDeviceDialogState = RemoveDeviceDialogState.Visible(
                            client = client
                        )
                    )
                }
```

I know it's not a lot of lines of code but it's used several times and my pedantic inner voice tells me "maybe it's worth extracting" 😄

File: app/src/main/kotlin/com/wire/android/ui/authentication/create/common/CreateAccountBaseViewModel.kt [hunk 4]
```diff
@@ -324,6 +327,12 @@ private fun RegisterClientResult.Failure.toCodeError() = when (this) {
     RegisterClientResult.Failure.TooManyClients -> CreateAccountCodeViewState.CodeError.TooManyDevicesError
     RegisterClientResult.Failure.InvalidCredentials -> CreateAccountCodeViewState.CodeError.DialogError.InvalidEmailError
     is RegisterClientResult.Failure.Generic -> CreateAccountCodeViewState.CodeError.DialogError.GenericError(this.genericFailure)
+    RegisterClientResult.Failure.PasswordAuthRequired -> {
+        Log.wtf("RegisterClient", "wrong password when register client after creating a new account")
+        CreateAccountCodeViewState.CodeError.DialogError.GenericError(
+            CoreFailure.Unknown(IllegalStateException("wrong password when register client after creating a new account"))
```
[the remark this answers is not in the document]
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/ui/authentication/create/common/CreateAccountBaseViewModel.kt:333]: After creating an account, we use the same password that the user entered to register the client, so PasswordAuthRequired or InvalidCredentials will never occur, i can change both to `CreateAccountCodeViewState.CodeError.None` and add some meaningful logs.

[the remark this answers is not in the document]
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/ui/authentication/create/common/CreateAccountBaseViewModel.kt]: https://github.com/wireapp/wire-android-reloaded/pull/682/commits/66531c59e31dd0d9b7e44f8687096e4dc19ff4ae

**Comment by AndroidBob** [comment]
Build [12](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/12/) **failed**.

**Comment by AndroidBob** [comment]
Build [13](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/13/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/authentication/create/common/CreateAccountBaseViewModel.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/authentication/create/common/CreateAccountBaseViewModel.kt]: I always have hard time trying to decide whether an error is worth crashing the app or not 😅 Do we have an option to send the logs somewhere to have the same output just like when the app crashes? If not and if this error shouldn't happen at all under normal circumstances then maybe it's ok to just crash the app 🤔

**Comment by AndroidBob** [comment]
Build [14](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/14/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3683369)(3683369)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3683369)(3683369)-dev-release.apk)

**Comment by AndroidBob** [comment]
Build [15](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/15/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3685091)(3685091)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3685091)(3685091)-dev-release.apk)

File: app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceScreen.kt [hunk 5]
```diff
@@ -75,7 +77,7 @@ private fun RemoveDeviceContent(
                 is RemoveDeviceState.Success ->
                     RemoveDeviceItemsList(lazyListState, state.deviceList, false, onItemClicked)
                 RemoveDeviceState.Loading ->
-                    RemoveDeviceItemsList(lazyListState, List(10) { Device() }, true, onItemClicked)
```
Comment (saleniuk) [inline app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceScreen.kt:78]: This `List(10) { Device() }` was to show the shimmer effect while loading data

Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceScreen.kt:78]: ohhh nice catch, i will add it again

**Comment by AndroidBob** [comment]
Build [25](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/25/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceScreen.kt (hunk 5 above)
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceScreen.kt]: @saleniuk Done!

**Comment by AndroidBob** [comment]
Build [27](https://10.10.124.134/job/wire-android-reloaded/job/PR-682/27/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3686196)(3686196)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3686196)(3686196)-dev-release.apk)

**Pull request merged**